cx/op_json.go

```
@@ -0,0 +1,175 @@
+			tokenType = JSON_TOKEN_DELIM
+			jsons[index].tokenDelim = value
+		case bool:
+			tokenType = JSON_TOKEN_BOOL
+			jsons[index].tokenBool = value
+		case float64:
+			tokenType = JSON_TOKEN_F64
+			jsons[index].tokenF64 = value
+		case json.Number:
+			tokenType = JSON_TOKEN_NUMBER
+			jsons[index].tokenNumber = value
+		case string:
+			tokenType = JSON_TOKEN_STR
+			jsons[index].tokenStr = value
+		default:
+			tokenType = JSON_TOKEN_NULL
+		}
+	}
+	jsons[index].tokenType = tokenType
+}
+
+// Type returns the type of the current token.
+func opJSONTokenType(expr *CXExpression, fp int) {
+	jsonFile := jsons[ReadI32(fp, expr.Inputs[0])]
+	WriteMemory(GetFinalOffset(fp, expr.Outputs[0]), FromI32(jsonFile.tokenType))
+}
+
+// Delim returns current token as an int32 delimiter.
+// Panics if token type is not JSON_TOKEN_DELIM.
+func opJSONTokenDelim(expr *CXExpression, fp int) {
+	jsonFile := jsons[ReadI32(fp, expr.Inputs[0])]
+	if jsonFile.tokenType != JSON_TOKEN_DELIM {
+		panic("json : not a delim value")
+	}
+	WriteMemory(GetFinalOffset(fp, expr.Outputs[0]), FromI32(int32(jsonFile.tokenDelim)))
+}
+
+// Bool returns current token as a bool value.
+// Panics if token type is not JSON_TOKEN_BOOL.
+func opJSONTokenBool(expr *CXExpression, fp int) {
+	jsonFile := jsons[ReadI32(fp, expr.Inputs[0])]
+	if jsonFile.tokenType != JSON_TOKEN_BOOL {
+		panic("json : not a bool value")
+	}
+	WriteMemory(GetFinalOffset(fp, expr.Outputs[0]), FromBool(jsonFile.tokenBool))
+}
+
+// Float64 returns current token as float64 value.
+// Panics if token can't be interpreted as float64 value.
+func opJSONTokenF64(expr *CXExpression, fp int) {
+	jsonFile := jsons[ReadI32(fp, expr.Inputs[0])]
+	var value float64
+	if jsonFile.tokenType == JSON_TOKEN_F64 {
+		value = jsonFile.tokenF64
+	} else if jsonFile.tokenType == JSON_TOKEN_NUMBER {
+		var err error
+		value, err = jsonFile.tokenNumber.Float64()
+		if err != nil {
+			panic(err)
+		}
+	} else {
+		panic("json : not a f64 value")
```

A panic is not a good way to handle an error in an interpreter unless it's an internal error.  Why not return an error in addition to the value?

---

Check can be done in cx side using the json.Type() function, it defers the decision to the user to checks if type conversion is valid or not. But you're right this is not a good way to handle error in an interpreter, I'll change it.